### PR TITLE
[ML-26956] Add support for Github Enterprise as CI/CD provider

### PR DESCRIPTION
**Pull request by zhe-db**
Added support for GitHub Enterprise as CI/CI provider

- Add an option for CICD provider when users create a project, now `GItHub Actions`, `GitHub Enterprise Actions`, `AzureDevOps`.
- Change `git-provider` to `gitHubEnterprise` payload for run-notebook github action according to [here](https://github.com/databricks/run-notebook/pull/29).
- Change `git-provider` payload to `gitHubEnterprise` for Databricks CLI to run notebooks according to  [here](https://src.dev.databricks.com/databricks/universe/-/blob/jobs-api/api/openapi/jobs-2.1.openapi.jsonnet?L1204).
- Change CD URL to `http(s)://HOSTNAME/api/v3/repos/OWNER/REPO/actions/workflows/WORKFLOW_ID/dispatches ` for gitHubEnterprise according to user supplied git repo URL according to github [docs](https://docs.github.com/en/enterprise-server@3.7/rest/actions/workflows).
- Add and fix unit tests.
- Testing with `mlops-azure-cuj`: https://github.com/databricks/mlops-azure-cuj/pull/60
- Testing workflow https://adb-3253686410953403.3.azuredatabricks.net/?o=3253686410953403#job/874940478032607/run/1231301

### CODE REVIEW

File: {{cookiecutter.project_name}}/notebooks/TriggerModelDeploy.py [hunk 1]
```diff
@@ -45,15 +45,24 @@
...
 )
-cd_trigger_url = f"https://api.github.com/repos/{github_repo}/actions/workflows/deploy-model-{env}.yml/dispatches"
+if github_server == "https://github.com" or github_server == "github.com":
+    github_server = "https://api.github.com"
+else:
+    # URL for github Enterprise Server: https://docs.github.com/en/enterprise-server@3.8/rest/actions/workflows
+    github_server = f"{github_server}/api/v3"
+cd_trigger_url = f"{github_server}/repos/{github_repo}/actions/workflows/deploy-model-{env}.yml/dispatches"
 authorization = f"token {token}"
+print(cd_trigger_url)
```
Comment (zhe-db) [inline {{cookiecutter.project_name}}/notebooks/TriggerModelDeploy.py:65]: used it for testing for now, will remove

File: README.md [hunk 2]
```diff
@@ -50,7 +50,7 @@ To create a new project, run:
 This will prompt for parameters for project initialization. Some of these parameters are required to get started:
  * ``project_name``: name of the current project
  * ``cloud``: Cloud provider you use with Databricks (AWS, Azure, or GCP)
- * ``cicd_platform`` : CI/CD platform of choice (GitHub Actions or Azure DevOps)
+ * ``cicd_platform`` : CI/CD platform of choice (GitHub Actions or Github Enterprise Actions or Azure DevOps)
```
Comment (mingyu89) [inline README.md:53]: How about (Github Actions, Github Actions for enterprises or Azure DevOps)? 
The "cicd_platform" variable needs to be changed accordingly.

I wasn't able to find term "Github Enterprise Actions".
Github doc
https://docs.github.com/en/enterprise-server@3.8/admin/github-actions/getting-started-with-github-actions-for-your-enterprise/about-github-actions-for-enterprises

File: hooks/post_gen_project.py [hunk 3]
```diff
@@ -55,7 +55,7 @@ def remove_filepath(filepath):
 
 
 cicd_specific_paths = {
-    "gitHub": [
+    "git": [
```
Comment (mingyu89) [inline hooks/post_gen_project.py:58]: Could you explain purpose of the changes in this file?

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md [hunk 4]
```diff
@@ -5,7 +5,9 @@ This directory contains setup scripts intended to automate CI/CD and ML resource
 for MLOps engineers.
 
 {% if cookiecutter.cicd_platform == "gitHub" -%}
-The scripts set up CI/CD with GitHub Actions. If using another CI/CD provider, you can
+The scripts set up CI/CD with GitHub Actions. 
+{% elif cookiecutter.cicd_platform == "gitHubEnterprise" -%}
```
Comment (mingyu89) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md:9]: Lines 10 to 20 are previous under "gitHub". After the change, they will be only under "gitHubEnterprise" and missing from "gitHub".

File: hooks/post_gen_project.py (hunk 3 above)
Comment (zhe-db) [reply hooks/post_gen_project.py:58]: `gitHub` and `gitHubEnterprise` both uses `.github` directory, that python cannot hash lists or sets as keys so I make a common key for both `gitHub` and `gitHubEnterprise`, the reason not to hash both `gitHub` and `gitHubEnterprise` is because of the below code blocks where we remove file paths of unwanted directories.

Comment (mingyu89) [reply hooks/post_gen_project.py:58]: Got it. It's the unrelated cleanup logic after the project generation. You are matching the prefix to map `gitHub` and `gitHubEnterprise` to the same entity in the map.

Since there are only 2 options, how about Removing the map and explicitly checking the condition
`cicd_platform == "gitHub"/"gitHubEnterprise"/"azureDevOpsServices"` and remove related files?

In this way there's no implicit cicd_platform name assumption.

Comment (zhe-db) [reply hooks/post_gen_project.py:58]: Agree, the remove_map logic is kinda weird, I will change to explicitly check condition.

File: hooks/post_gen_project.py [hunk 5]
```diff
@@ -53,21 +53,11 @@ def remove_filepath(filepath):
...
-        os.path.join(".azure"),
-        os.path.join(".mlops-setup-scripts", "cicd", "azure-devops.tf"),
-    ],
-}
-
-for cicd, paths in cicd_specific_paths.items():
-    if cicd != cicd_platform:
-        for path in paths:
-            remove_filepath(path)
+if cicd_platform in ["gitHub", "gitHubeEnterprise"]:
```
Comment (arpitjasa-db) [inline hooks/post_gen_project.py:56]:
```suggestion
if cicd_platform in ["gitHub", "gitHubEnterprise"]:
```

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md [hunk 6]
```diff
@@ -198,6 +200,9 @@ python .mlops-setup-scripts/cicd/bootstrap.py \
 {%- if cookiecutter.cicd_platform == "gitHub" %}
   --var github_repo_url=https://github.com/<your-org>/<your-repo-name> \
   --var git_token=<your-git-token>
+{%- elif cookiecutter.cicd_platform == "gitHubEnterprise" %}
+   -- var github_repo_url=<your-enterprise-github-server>/<your-org>/<your-repo-name> \
+   -- var git_token=<your-git-token>
```
Comment (arpitjasa-db) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md:205]:
```suggestion
   --var github_repo_url=<your-enterprise-github-server>/<your-org>/<your-repo-name> \
   --var git_token=<your-git-token>
```
Also would it be `github_repo_url` or `github_enterprise_repo_url`?

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf [hunk 7]
```diff
@@ -19,10 +19,18 @@ resource "databricks_secret" "cd_credentials_token" {
   scope        = databricks_secret_scope.cd_credentials.id
 }
 
+resource "databricks_secret" "cd_github_server" {
+  key = "github_server"
+  // regex expression for parsing github server url out of github repo url
+  // example: http(s)://mygitserver.anything/org/repo -> http(s)://mygitserver.anything
+  string_value = regex(".*\/\/[^\/]+", var.github_repo_url)
+  scope = databricks_secrete_scope.cd_credentials.id
```
Comment (arpitjasa-db) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf:27]:
```suggestion
  scope = databricks_secret_scope.cd_credentials.id
```

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf [hunk 8]
```diff
@@ -19,10 +19,18 @@ resource "databricks_secret" "cd_credentials_token" {
...
+}
+
 resource "databricks_secret" "cd_github_repo" {
   key = "github_repo"
   // Extract the 'organization/repo' substring used to identify the repo
-  string_value = replace(var.github_repo_url, "https://github.com/", "")
+  string_value = replace(var.github_repo_url, databricks_secret.cd_github_server, "")
```
Comment (arpitjasa-db) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf:33]: I think you'd need to do `databricks_secret.cd_github_server.value` but that also may not work for sensitivity. Could be easier to just pull out the regex logic into a Terraform `local` and pass it into both `resources`.

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf [hunk 9]
```diff
@@ -19,10 +19,18 @@ resource "databricks_secret" "cd_credentials_token" {
   scope        = databricks_secret_scope.cd_credentials.id
 }
 
+resource "databricks_secret" "cd_github_server" {
+  key = "github_server"
+  // regex expression for parsing github server url out of github repo url
+  // example: http(s)://mygitserver.anything/org/repo -> http(s)://mygitserver.anything
+  string_value = regex(".*\/\/[^\/]+", var.github_repo_url)
```
Comment (arpitjasa-db) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf:26]: This regex scares me a bit, can we confirm this works as expected via some tests? Wonder if it would be simpler to ask them to provide it

File: {{cookiecutter.project_name}}/notebooks/TriggerModelDeploy.py [hunk 10]
```diff
@@ -45,15 +45,24 @@
...
+    f"{env}-{{cookiecutter.project_name}}-cd-credentials", "github_server"
+)
 github_repo = dbutils.secrets.get(
     f"{env}-{{cookiecutter.project_name}}-cd-credentials", "github_repo"
 )
 token = dbutils.secrets.get(
     f"{env}-{{cookiecutter.project_name}}-cd-credentials", "token"
 )
-cd_trigger_url = f"https://api.github.com/repos/{github_repo}/actions/workflows/deploy-model-{env}.yml/dispatches"
+if github_server == "https://github.com" or github_server == "github.com":
```
Comment (arpitjasa-db) [inline {{cookiecutter.project_name}}/notebooks/TriggerModelDeploy.py:58]: Might be safer to use `cookiecutter.cicd_platform` here as well instead of bifurcating on URLs (someone could use a `http` or throw in a `www`)

**Review by arpitjasa-db — COMMENTED** [review]
Awesome work @zhe-db! Left a few minor comments

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md (hunk 6 above)
Comment (zhe-db) [reply {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md:205]: I use the same variable for `gitHub` and `gitHubEnterprise` to reduce complexity since we only allow one choice.

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf (hunk 8 above)
Comment (zhe-db) [reply {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf:33]: Cool, sounds good.

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf (hunk 9 above)
Comment (zhe-db) [reply {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf:26]: Cool, will add another variable for github_server

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md [hunk 11]
```diff
@@ -198,6 +211,10 @@ python .mlops-setup-scripts/cicd/bootstrap.py \
 {%- if cookiecutter.cicd_platform == "gitHub" %}
   --var github_repo_url=https://github.com/<your-org>/<your-repo-name> \
   --var git_token=<your-git-token>
+{%- elif cookiecutter.cicd_platform == "gitHubEnterprise" %}
+   --var github_repo_url=<your-enterprise-github-server>/<your-org>/<your-repo-name> \
+   --var git_token=<your-git-token>
```
Comment (zhe-db) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md:216]: missing a `\`, will add

File: {{cookiecutter.project_name}}/.github/workflows/run-tests-fs.yml [hunk 12]
```diff
@@ -148,7 +148,7 @@ jobs:
             ],
             "git_source": {
               "git_url": {% raw %}"${{ github.server_url }}/${{ github.repository }}",{% endraw %}
-              "git_provider": "gitHub",
+              "git_provider": "{{cookiecutter.cicd_platform}}",
```
Comment (vladimirk-db) [inline {{cookiecutter.project_name}}/.github/workflows/run-tests-fs.yml:151]: Can we verify diffs for the gitHub option as well (that there's no regression).

You can do two separate commits of a generated Stacks project and then diff them.

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf [hunk 13]
```diff
@@ -19,10 +19,16 @@ resource "databricks_secret" "cd_credentials_token" {
   scope        = databricks_secret_scope.cd_credentials.id
 }
 
+resource "databricks_secret" "cd_github_server" {
```
Comment (vladimirk-db) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf:22]: Wait, is the cd_github_server a databricks secret? with a scope?  It's not a credential, right?

Comment (vladimirk-db) [reply {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf:22]: (Maybe I'm missing something.)

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf [hunk 14]
```diff
@@ -19,10 +19,16 @@ resource "databricks_secret" "cd_credentials_token" {
...
+  key = "github_server"
+  string_value = var.github_server_url
+  scope = databricks_secret_scope.cd_credentials.id
+}
+
 resource "databricks_secret" "cd_github_repo" {
```
Comment (vladimirk-db) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf:28]: Same thing here - why is the github_repo a databricks_secret?

File: README.md (hunk 2 above)
Comment (vladimirk-db) [reply README.md:53]: I wouldn't do Github Enterprise Actions... "Github Actions (Enterprise)" or "Github Actions for Enterprises" maybe? The latter seems to be the official name.

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf (hunk 13 above)
Comment (zhe-db) [reply {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf:22]: I am following the same style for `github_repo` [here](https://src.dev.databricks.com/databricks/mlops-stack/-/blob/%7B%7Bcookiecutter.project_name%7D%7D/.mlops-setup-scripts/cicd/common/main.tf?L22). I think the reason is to facilitate retrieving the value through dbutils for example [here](https://src.dev.databricks.com/databricks/mlops-stack/-/blob/%7B%7Bcookiecutter.project_name%7D%7D/notebooks/TriggerModelDeploy.py?L49).

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf (hunk 14 above)
Comment (zhe-db) [reply {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf:28]: Please see above and correct me if I explained it wrong @arpitjasa-db

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf (hunk 13 above)
Comment (vladimirk-db) [reply {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf:22]: Ack, please double check this w/ @arpitjasa-db .

File: README.md (hunk 2 above)
Comment (arpitjasa-db) [reply README.md]: Looks like the official name is [GitHub Actions for GitHub Enterprise Server](https://docs.github.com/en/enterprise-server@3.5/admin/github-actions/getting-started-with-github-actions-for-your-enterprise/getting-started-with-github-actions-for-github-enterprise-server)

File: {{cookiecutter.project_name}}/notebooks/TriggerModelDeploy.py [hunk 15]
```diff
@@ -45,15 +45,27 @@
 assert env != "None", "env notebook parameter must be specified"
 assert model_uri != "", "model_uri notebook parameter must be specified"
 
-{% if cookiecutter.cicd_platform == "gitHub" -%}
+{% if cookiecutter.cicd_platform in ["gitHub", "gitHubEnterprise"] -%}
+github_server = dbutils.secrets.get(
```
Comment (arpitjasa-db) [inline {{cookiecutter.project_name}}/notebooks/TriggerModelDeploy.py:49]: We can move this logic inside the `elif` on L61 since we only use the `github_server` value in the enterprise case

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/variables.tf [hunk 16]
```diff
@@ -9,14 +9,20 @@ variable "git_provider" {
...
+  description = "URL of the hosted git server containing the current ML project, e.g. https://github.com/"
+  default     = "https://github.com/"
+}
+
 variable "env" {
   type        = string
   description = "Current env, i.e. 'staging' or 'prod'"
 }
-{%- endif -%}
+{%- endif -%}
```
Comment (arpitjasa-db) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/variables.tf:28]: nit newline

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf (hunk 13 above)
Comment (arpitjasa-db) [reply {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf:22]: Yeah we could consider instead writing out to a `json` output for example and pulling the value in from there for non-sensitive values, but this is fine for now since it's following the preset convention

Comment (vladimirk-db) [reply {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/common/main.tf:22]: The issue is that it's misleading - for instance the scope value does not apply, but it's present. Can we at least document this and then follow-up (don't need to block this PR on it).

File: README.md (hunk 2 above)
Comment (vladimirk-db) [reply README.md]: Damn, that's a mouthful but it's better (can't argue with official).

**Review by vladimirk-db — APPROVED** [review]
LGTM @zhe-db !
